content/oci/oci.go

```diff
@@ -0,0 +1,227 @@
+	}
+
+	return store, nil
+}
+
+// Fetch fetches the content identified by the descriptor.
+func (s *Store) Fetch(ctx context.Context, target ocispec.Descriptor) (io.ReadCloser, error) {
+	return s.storage.Fetch(ctx, target)
+}
+
+// Push pushes the content, matching the expected descriptor.
+func (s *Store) Push(ctx context.Context, expected ocispec.Descriptor, reader io.Reader) error {
+	if err := s.storage.Push(ctx, expected, reader); err != nil {
+		return err
+	}
+
+	return s.finder.Index(ctx, s.storage, expected)
+}
+
+// Exists returns true if the described content exists.
+func (s *Store) Exists(ctx context.Context, target ocispec.Descriptor) (bool, error) {
+	return s.storage.Exists(ctx, target)
+}
+
+// Tag tags a descriptor with a reference string.
+// Returns ErrUnsupportedMediaType if the media type is not supported for tagging.
+func (s *Store) Tag(ctx context.Context, desc ocispec.Descriptor, reference string) error {
+	if reference == "" {
+		return errdef.ErrInvalidReference
+	}
+
+	exists, err := s.storage.Exists(ctx, desc)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return fmt.Errorf("%s: %s: %w", desc.Digest, desc.MediaType, errdef.ErrNotFound)
+	}
+
+	// reference can be either a tag or a digest.
+	switch desc.MediaType {
+	case ocispec.MediaTypeImageManifest, ocispec.MediaTypeImageIndex:
+		if !isDigest(reference) {
+			if desc.Annotations == nil {
+				desc.Annotations = map[string]string{}
+			}
+			desc.Annotations[ocispec.AnnotationRefName] = reference
+		}
+
+		s.resolver.Tag(ctx, desc, reference)
+		return s.saveIndex()
+	default:
+		return fmt.Errorf("%s: %s: %w", desc.Digest, desc.MediaType, errdef.ErrUnsupportedMediaType)
+	}
+}
+
+// Resolve resolves a reference to a descriptor.
+func (s *Store) Resolve(ctx context.Context, reference string) (ocispec.Descriptor, error) {
+	if reference == "" {
+		return ocispec.Descriptor{}, errdef.ErrInvalidReference
+	}
+
+	return s.resolver.Resolve(ctx, reference)
+}
+
+func (s *Store) UpEdges(ctx context.Context, node ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+	return s.finder.UpEdges(ctx, node)
+}
+
+// validateOCILayoutFile ensures the `oci-layout` file.
+func (s *Store) validateOCILayoutFile() error {
+	layoutFilePath := filepath.Join(s.root, ocispec.ImageLayoutFile)
+	layoutFile, err := os.Open(layoutFilePath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		}
+
+		layout := ocispec.ImageLayout{
+			Version: ocispec.ImageLayoutVersion,
+		}
+		layoutJSON, err := json.Marshal(layout)
+		if err != nil {
+			return err
+		}
+
+		return ioutil.WriteFile(layoutFilePath, layoutJSON, 0666)
+	}
+	defer layoutFile.Close()
+
+	var layout *ocispec.ImageLayout
+	err = json.NewDecoder(layoutFile).Decode(&layout)
+	if err != nil {
+		return err
+	}
+	if layout.Version != ocispec.ImageLayoutVersion {
+		return errdef.ErrUnsupportedVersion
+	}
+
+	return nil
+}
+
+// load reads the index.json from the file system.
+func (s *Store) loadIndex() error {
```

Comment:
Just to curious as to why `loadIndex` and 'saveIndex` are not being exported anymore? If I remember correctly, the other OCI target had it exported and it could be used outside.

Reply:
Do we need to manually call `loadIndex` and `saveIndex` somewhere outside? The OCI does it automatically when tagging and resolving.

Reply:
@shizhMSFT Yeah it makes sense that we provide both ways and users can decide to auto save the index or not. I just exported the `SaveIndex()` method, and added a `AutoSaveIndex` option in the OCI store, and make it default to true. The users can set it to false if they want to manually save the index like:
```go
s := oci.New(root)
s.AutoSaveIndex = false
// do something

s.Tag(foo)
s.Tag(bar)

// now save the index
s.SaveIndex()
```